File: agent-service/src/services/repoExplorer.ts
import * as fs from "fs";
import * as path from "path";
import * as childProcess from "child_process";
import simpleGit from "simple-git";
import { config } from "../core/config";
import { decrypt } from "../core/crypto";

// Repos are cloned to a dedicated "planning" sub-workspace so they don't
// interfere with execution workspaces.
const PLANNING_ROOT = path.join(config.WORKSPACES_ROOT, "planning");

const IGNORE_DIRS = new Set([
  ".git", "node_modules", "__pycache__", ".next", "dist", "build",
  ".venv", "venv", "vendor", ".turbo", "coverage", ".mypy_cache",
  ".pytest_cache", ".cache",
]);

export interface RepoRef {
  id: string;
  name: string;
  remote_url?: string;
  branch?: string;
  auth_type?: string;
  credentials?: string;
  local_path?: string;
}

// ─── Clone / pull ─────────────────────────────────────────────────────────────

/**
 * Returns a local path where the repo is available.
 * If already cloned (by execution scheduler), reuses that path.
 * Otherwise does a shallow clone into the planning workspace.
 */
export async function ensureLocalPath(repo: RepoRef, projectId: string): Promise<string> {
  // Prefer execution workspace path if already cloned
  if (repo.local_path && fs.existsSync(repo.local_path)) {
    return repo.local_path;
  }

  const planningPath = path.join(PLANNING_ROOT, projectId, repo.name);

  if (fs.existsSync(planningPath)) {
    // Already cloned in planning workspace — try to pull latest
    try { await simpleGit(planningPath).pull(); } catch { /* ignore network errors */ }
    return planningPath;
  }

  if (!repo.remote_url) {
    throw new Error(`Repository "${repo.name}" has no remote URL and is not yet cloned`);
  }

  fs.mkdirSync(path.dirname(planningPath), { recursive: true });

  let cloneUrl = repo.remote_url;
  if (repo.auth_type === "token" && repo.credentials) {
    const token = decrypt(repo.credentials);
    const url = new URL(cloneUrl);
    url.username = token;
    cloneUrl = url.toString();
  }

  await simpleGit().clone(cloneUrl, planningPath, [
    "--depth=1",
    "--branch", repo.branch ?? "main",
    "--single-branch",
  ]);

  return planningPath;
}

// ─── Filesystem tools ─────────────────────────────────────────────────────────

/** Lists files and directories at `relPath` inside the repo. */
export function listDirectory(localPath: string, relPath = ""): string {
  const target = relPath ? path.join(localPath, relPath) : localPath;

  if (!isInsideRoot(localPath, target)) return "Error: path traversal not allowed";

  try {
    if (!fs.existsSync(target)) return `Path not found: ${relPath || "/"}`;
    if (!fs.statSync(target).isDirectory()) return `"${relPath}" is a file, not a directory`;

    const entries = fs.readdirSync(target, { withFileTypes: true });
    const filtered = entries.filter((e) => !IGNORE_DIRS.has(e.name) && !e.name.startsWith("."));
    filtered.sort((a, b) => {
      if (a.isDirectory() !== b.isDirectory()) return a.isDirectory() ? -1 : 1;
      return a.name.localeCompare(b.name);
    });

    if (filtered.length === 0) return "(empty directory)";

    return filtered
      .map((e) => `${e.isDirectory() ? "📁" : "📄"} ${e.name}${e.isDirectory() ? "/" : ""}`)
      .join("\n");
  } catch (e) {
    return `Error listing directory: ${e}`;
  }
}

/** Reads file content at `relPath` inside the repo. Truncates large files. */
export function readFile(localPath: string, relPath: string, maxChars = 10_000): string {
  const target = path.join(localPath, relPath);

  if (!isInsideRoot(localPath, target)) return "Error: path traversal not allowed";

  try {
    if (!fs.existsSync(target)) return `File not found: ${relPath}`;
    if (fs.statSync(target).isDirectory()) return `"${relPath}" is a directory — use list_directory instead`;

    const raw = fs.readFileSync(target, "utf-8");
    if (raw.length <= maxChars) return raw;
    return raw.slice(0, maxChars) + `\n\n[... truncated — ${raw.length - maxChars} more characters]`;
  } catch (e) {
    return `Error reading file: ${e}`;
  }
}

/**
 * Searches for a text pattern across the repo (like grep -r).
 * Returns matching file paths and line excerpts (up to 30 matches).
 */
export function searchCode(localPath: string, pattern: string, fileGlob = "*"): string {
  try {
    const safePattern = pattern.replace(/'/g, "'\\''");
    const result = childProcess.execSync(
      `grep -r --include="${fileGlob}" -n "${safePattern}" . 2>/dev/null | head -30`,
      { cwd: localPath, timeout: 8000, encoding: "utf-8" }
    );
    return result.trim() || "No matches found";
  } catch {
    return "No matches found";
  }
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

function isInsideRoot(root: string, target: string): boolean {
  const resolvedRoot = path.resolve(root);
  const resolvedTarget = path.resolve(target);
  return resolvedTarget.startsWith(resolvedRoot + path.sep) || resolvedTarget === resolvedRoot;
}

File: agent-service-py/src/agent_service/services/repo_explorer.py
from __future__ import annotations

import asyncio
import logging
import os
import subprocess
from pathlib import Path
from typing import Optional
from urllib.parse import urlparse

from agent_service.config import config
from agent_service.crypto import decrypt
from agent_service.services import data_client

logger = logging.getLogger(__name__)

PLANNING_ROOT = os.path.join(config.WORKSPACES_ROOT, "planning")

IGNORE_DIRS = frozenset(
    {
        ".git",
        "node_modules",
        "__pycache__",
        ".next",
        "dist",
        "build",
        ".venv",
        "venv",
        "vendor",
        ".turbo",
        "coverage",
        ".mypy_cache",
        ".pytest_cache",
        ".cache",
    }
)


async def ensure_local_path(repo: dict, project_id: str) -> str:
    """
    Return a local filesystem path where the repo is available.
    Reuses the execution workspace path if already cloned; otherwise
    does a shallow clone into the planning workspace.
    """
    local_path: Optional[str] = repo.get("local_path")
    if local_path and os.path.exists(local_path):
        return local_path

    planning_path = os.path.join(PLANNING_ROOT, project_id, repo["name"])

    if os.path.exists(planning_path):
        # Already cloned — try to pull latest
        try:
            proc = await asyncio.create_subprocess_exec(
                "git", "-C", planning_path, "pull",
                stdout=asyncio.subprocess.DEVNULL,
                stderr=asyncio.subprocess.DEVNULL,
            )
            await proc.wait()
        except Exception:
            pass
        return planning_path

    remote_url = repo.get("remote_url")
    if not remote_url:
        raise ValueError(
            f'Repository "{repo["name"]}" has no remote URL and is not yet cloned'
        )

    os.makedirs(os.path.dirname(planning_path), exist_ok=True)

    clone_url = remote_url
    if repo.get("auth_type") == "token" and repo.get("credentials"):
        token = decrypt(repo["credentials"])
        parsed = urlparse(clone_url)
        clone_url = parsed._replace(
            netloc=f"{token}@{parsed.hostname or ''}{f':{parsed.port}' if parsed.port else ''}"
        ).geturl()

    branch = repo.get("branch") or "main"

    proc = await asyncio.create_subprocess_exec(
        "git", "clone", clone_url, planning_path,
        "--depth=1",
        f"--branch={branch}",
        "--single-branch",
        stdout=asyncio.subprocess.DEVNULL,
        stderr=asyncio.subprocess.PIPE,
    )
    _, stderr = await proc.communicate()
    if proc.returncode != 0:
        raise RuntimeError(
            f"git clone failed for {repo['name']}: {stderr.decode(errors='replace')}"
        )

    # Update repo record with local path
    await data_client.update_repository_status(repo["id"], {"status": "ready", "local_path": planning_path})
    return planning_path


def list_directory(local_path: str, rel_path: str = "") -> str:
    """List files and directories at ``rel_path`` inside the repo."""
    target = os.path.join(local_path, rel_path) if rel_path else local_path

    if not _is_inside_root(local_path, target):
        return "Error: path traversal not allowed"

    try:
        if not os.path.exists(target):
            return f"Path not found: {rel_path or '/'}"
        if not os.path.isdir(target):
            return f'"{rel_path}" is a file, not a directory'

        entries = os.scandir(target)
        filtered = sorted(
            [e for e in entries if e.name not in IGNORE_DIRS and not e.name.startswith(".")],
            key=lambda e: (not e.is_dir(), e.name.lower()),
        )

        if not filtered:
            return "(empty directory)"

        lines: list[str] = []
        for entry in filtered:
            suffix = "/" if entry.is_dir() else ""
            icon = "📁" if entry.is_dir() else "📄"
            lines.append(f"{icon} {entry.name}{suffix}")
        return "\n".join(lines)
    except Exception as exc:
        return f"Error listing directory: {exc}"


def read_file(local_path: str, rel_path: str, max_chars: int = 10_000) -> str:
    """Read file content at ``rel_path`` inside the repo. Truncates large files."""
    target = os.path.join(local_path, rel_path)

    if not _is_inside_root(local_path, target):
        return "Error: path traversal not allowed"

    try:
        if not os.path.exists(target):
            return f"File not found: {rel_path}"
        if os.path.isdir(target):
            return f'"{rel_path}" is a directory — use list_directory instead'

        with open(target, "r", encoding="utf-8", errors="replace") as fh:
            raw = fh.read()

        if len(raw) <= max_chars:
            return raw
        return raw[:max_chars] + f"\n\n[... truncated — {len(raw) - max_chars} more characters]"
    except Exception as exc:
        return f"Error reading file: {exc}"


def search_code(local_path: str, pattern: str, file_glob: str = "*") -> str:
    """Search for a text pattern across the repo (like grep -r)."""
    try:
        result = subprocess.run(
            ["grep", "-r", f"--include={file_glob}", "-n", pattern, "."],
            cwd=local_path,
            capture_output=True,
            text=True,
            timeout=8,
        )
        lines = result.stdout.strip().split("\n")[:30]
        filtered = [l for l in lines if l]
        return "\n".join(filtered) if filtered else "No matches found"
    except Exception:
        return "No matches found"


def _is_inside_root(root: str, target: str) -> bool:
    resolved_root = os.path.realpath(root)
    resolved_target = os.path.realpath(target)
    return resolved_target == resolved_root or resolved_target.startswith(resolved_root + os.sep)
